utils/cipher.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "cipher.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include "third_party/mbedtls/include/mbedtls/aes.h"
+#include "third_party/mbedtls/include/mbedtls/md.h"
+
+#define MAX_TIMESTAMP_LEN 20
+
+int aes_encrypt(const char *plaintext, int plaintext_len, const unsigned char *key, unsigned int keybits,
+                unsigned char iv[AES_BLOCK_SIZE], char *ciphertext, int ciphertext_len) {
+  mbedtls_aes_context ctx;
+  int status;
+  unsigned char buf[AES_BLOCK_SIZE];
+  int n = 0;
+  char *err;
+  /*
+   * Initialise the encryption operation.
+   */
+  // Check ciphertext has enough space
+  int new_len = plaintext_len + (AES_BLOCK_SIZE - plaintext_len % 16);
+  if (new_len > ciphertext_len) {
+    err = "ciphertext has not enough space";
+    goto exit;
+  }
+  mbedtls_aes_init(&ctx);
+  memset(ciphertext, 0, ciphertext_len);
+  /* set encryption key */
+  if ((status = mbedtls_aes_setkey_enc(&ctx, key, keybits)) != 0) {
+    err = "set aes key failed";
+    goto exit;
+  }
+
+  /*
+   * Encrypt plaintext
+   */
+  for (int i = 0; i < new_len; i += AES_BLOCK_SIZE) {
+    memset(buf, 0, AES_BLOCK_SIZE);
+    n = (new_len - i > AES_BLOCK_SIZE) ? 16 : (int)(new_len - i);
+    memcpy(buf, plaintext + i, n);
+    if ((status = mbedtls_aes_crypt_cbc(&ctx, MBEDTLS_AES_ENCRYPT, AES_BLOCK_SIZE, iv, buf, buf)) != 0) {
+      err = "aes decrpyt failed";
+      goto exit;
+    }
+    memcpy(ciphertext, buf, AES_BLOCK_SIZE);
+    ciphertext += AES_BLOCK_SIZE;
+  }
+
+  /* Clean up */
+  mbedtls_aes_free(&ctx);
+  return new_len;
+exit:
+  fprintf(stderr, "%s\n", err);
+  mbedtls_aes_free(&ctx);
+  return -1;
+}
+
+endpoint_retcode_t aes_decrypt(const char *ciphertext, int ciphertext_len, const unsigned char *key,
```

Comment:
How about introducing context which contains all the parameters? So that, we don't have to pass all parameters to function `aes_decrypt`. Instead, compilers can generate better code for efficient parameter passing (all registers w/o stack).

Reply:
Great idea! I will do this after all unit tests were passed.